### PR TITLE
Made append_array function work with vectors of integers (more for is…

### PR DESCRIPTION
**Pull request by bbbales2**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

Two things (regarding https://github.com/stan-dev/math/issues/481),

I added a `vector<int> append_array(vector<int>, vector<int>)` signature. I think it makes sense to have at least that. I was putting together the Stan hooks for the append_array and it just felt like I needed the int[] signatures to make this complete. Maybe this should really be like rep_array and just accept all different types of things that can be put in a Stan array?

I changed mixed append_array tests to be combinations of double/fvar<var> and double/fvar<fvar<var>> instead of fvar<var>/fvar<fvar<var>>

~~(For some reason Github wouldn't let me upload to main Stan repos which is why this is living over here)~~ This got fixed. I should be able to put things in the regular Stan repos next time

#### How to Verify:

./runTests.py test/unit/math/prim/arr/fun/append_array_test
./runTests.py test/unit/math/mix/arr/fun/append_array_test

#### Side Effects:

Hopefully none

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): University of California Santa Barbara

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by seantalts** [comment]
Can you elaborate a little on your reasoning for adding this specialization? I'm not opposed I just didn't understand your comments above. Thanks!

**Comment by bbbales2** [comment]
Well most of the array reduction functions in the manual min/max stuff like that works with reals or ints. So I figured at bare minimum this should match that just to avoid annoying folks.

But given that rep_array supports all types (vectors, matrices, row_vectors, scalars), I think this sorta thing probably should as well (seems like they'd have similar applications). Looking at Rep array, I think that'd just be writing another version like this:

```
template <typename T>
inline std::vector<T>
append_array(const std::vector<T>& x, const std::vector<T>& y)
```

And adding a few simple tests with different objects moving around. Hopefully that template wouldn't confuse the type promoting version. I can add that if it seems reasonable, or just stick to the original plan.

**Comment by seantalts** [comment]
I agree that this should work for all types, and likely wouldn't need to handle any types individually.

I think there is already a generic version in the codebase (that I think you contributed), no? Is it missing something? I'm talking about the one here: https://github.com/stan-dev/math/blob/develop/stan/math/prim/arr/fun/append_array.hpp#L22

**Comment by bbbales2** [comment]
How does boost promote_args handle higher level types? Like if what came in were two vector<Matrix<var>>s, or maybe one vector<Matrix<var>> and one vector<Matrix<double>>?

**Comment by seantalts** [comment]
From what little I understand, promote_args relies on a vision of promotion as implemented by implicit conversion constructors. So if a T1 can be set to a T2, that's promotion. I think in those cases you mentioned it would work fine as long as the types are the same and might not work for something like `vector<Matrix<double>>` mixing with `vector<Matrix<var>>`. I'm not sure if the intention was to support that; the spec just says for two things of the same type.

Anyway, when in doubt test it out! A test for these use cases sounds necessary no matter the underlying implementation.

**Comment by bbbales2** [comment]
Sure @ the tests. No reason for them to be long.

The issue with not specializing is promote_args<int, int>::type is double (from http://www.boost.org/doc/libs/1_64_0/boost/math/tools/promotion.hpp) so this specialization is necessary I think. I saw int specializations in: https://github.com/stan-dev/math/blob/develop/stan/math/prim/mat/fun/min.hpp as well.

**Comment by seantalts** [comment]
Hm, I didn't realize promote_args would always convert int to double; now I see why you added the `int` specialization and how the other version you proposed above differs from the existing one. Sorry about that! 

Though, from the comments in the same promotion.hpp file: 
>       // But since most (all?) the math functions convert to double internally,

So I think it might be fine to promote to double? @bob-carpenter what do you think?

I think the int specialization in min that you've linked to in `min` is actually providing different behavior for int vs. double - for ints it will actually throw an exception if the list is empty, whereas for doubles it will return negative infinity (a concept not present in the binary representation of int). It's subtle and I'm not sure why double swallows the error instead of also throwing, but that seems to be the reason for the specialization.

**Comment by bbbales2** [comment]
I don't think promoting int to double is what we want. The use case I see for this is people messing with weird ragged arrays, in which case ints are indices and need to stay that way.

**Comment by seantalts** [comment]
Good point. In that case I'm partial to the version you posted above for type `T` - my thinking is that it will provide a less complicated route that's sure not to mutate any types where they're the same and should handle the int, int case. Thoughts?

**Comment by bob-carpenter** [comment]
As to https://github.com/stan-dev/math/pull/554#issuecomment-301346732, `boost::promote_args` is well documented by Boost.  For us, it promotes `int` and `double` to `double` and either of those and an autodiff type to the autodiff type, and that's it.  It is *not* covariant/contravariant as you might hope from languages like ML, so you don't get reasoning through the contained types.

**Comment by bob-carpenter** [comment]
For https://github.com/stan-dev/math/pull/554#issuecomment-301475602, I'd say that you can have a generic type for containers, but you need to deal with the `int` to `double` promotion explicitly.  

This is done really ineffectively throughout Stan.  In general, I think it would be good if we could make the `int` to `double` conversion covariant.  Right now, we allow `int` wherever we allow `real` (or at least try hard to do that).  But we don't allow `int[]` wherever we allow `real[]`.  If we can assign a `T1` to a `T2` then we should be able to assigna  container of `T1`s to a container of `T2`s.  

What I don't want to get in the habit of doing is mixing array and matrix types.  So I'm not in favor of appending integer arrays to vectors.  On the other hand, we have explicit conversion functions for a lot of that.

**Comment by seantalts** [comment]
So Bob, are you saying we do need an explicit `int` specialization here? Why wouldn't @bbbales2 other proposal (copied below) work to make sure we don't promote int to double? Here it is:

```
template <typename T>
inline std::vector<T>
append_array(const std::vector<T>& x, const std::vector<T>& y)
```

I would expect this to take precedence over the other implementation using `promote_args` any time the two types are the same, right? So I think in the case above where we know we don't want to promote two int arrays appended as an array of doubles, that'd handle it correctly right?

Sorry if I'm missing something! Still feel new to lots of this.

**Comment by bob-carpenter** [comment]
I didn't even look at the specific problem.

The matching happens based on the arguments, not the return type, so
promote_args won't be involved (you can't use it for argument type
inference, either).

That version will match vector<T> and vector<T> combos, but it won't
obvious allow you to append vector<double> and vector<int> into vector<double>.

Maybe we're not communicating on signatures.  I think it'd be nice to
have

  vector<double> append_array(const vector<double>& x, const vector<int>& y);

and the other way around.

**Comment by seantalts** [comment]
Wouldn't the [existing version](https://github.com/stan-dev/math/blob/develop/stan/math/prim/arr/fun/append_array.hpp#L22) take care of the "append vector<double> and vector<int> into vector<double>" use case?

And the one other one proposed above with just one `T` would handle the `int, int` case, right?

**Comment by bob-carpenter** [comment]
Yes, that will do it.  I can't think of a case where that won't do everything we need.   The reserve() and insert() idiom is as efficient as this can be, I think, no matter what the types are.

**Comment by bbbales2** [comment]
I did these things:

1. Added a template specialization for the matrix types
2. Moved everything into the mat directories
3. Left the int version of the function alone
4. Removed a lot of tests (I left in some random, easy to read spot checks on the derivatives)

The questions I have are:
1. Like the looks of the matrix version? If so I'll add some tiny tests in rev/fwd/mix. There's none now. Only a couple minor tests in prim. If no, I'll take it out and move everything back to array. No probs either way. Both things are really easy.

2. Tests shorter or longer? It's super easy for me to make them shorter.

With regards to our conversation, if I tried to get away with a function:

```
template <typename T>
inline std::vector<T>
append_array(const std::vector<T>& x, const std::vector<T>& y)
```

The compiler would get confused about what version of the function to use for Matrices (since strictly speaking, the Matrices are both of the same type and so can match that template). This doesn't work though, because matrices need to be separate code so I can spot check the sizes. So I just have one version for ints and one version for Matrices.

**Comment by bob-carpenter** [comment]
> On May 17, 2017, at 11:56 PM, Ben Bales <notifications@github.com> wrote:
> 
> I did these things:
> 
> 	• Added a template specialization for the matrix types
> 	• Moved everything into the mat directories
> 	• Left the int version of the function alone
> 	• Removed a lot of tests (I left in some random, easy to read spot checks on the derivatives)
> The questions I have are:
> 
> 	• Like the looks of the matrix version? If so I'll add some tiny tests in rev/fwd/mix. There's none now. Only a couple minor tests in prim. If no, I'll take it out and move everything back to array. No probs either way. Both things are really easy.
> 
> 	• Tests shorter or longer? It's super easy for me to make them shorter.
> 
> With regards to our conversation, if I tried to get away with a function:
> 
> template <typename T>
> inline std::vector<T>
> append_array(const std::vector<T>& x, const std::vector<T>& y)
> 
> The compiler would get confused about what version of the function to use for Matrices

Rule number one of the pragmatic programmer:  the compiler isn't broken.
The rules for matching functions are to find the best base template (or non-template)
overload, then choose the most appropriate template specialization if you chose a template
in the first step.  You can introduce ambiguities if there's not a single best match at
either stage.

I don't see why we'd ever want that append-array signature.  In general, we
want to be able to append vector<var> and vector<double>, and similarly on
down.  So we want generic inputs that can resolve promote_args for their base
scalars.  So we also want to be able to append vector<Matrix<var,-1,-1.> and 
vector<Matrix<double,-1,-1>> as well as append vector<vector<var>> and vector<vector<double>>.
There's no special logic for when the types happen to be the same.

> (since strictly speaking, the Matrices are both of the same type and so can match that template). This doesn't work though, because matrices need to be separate code so I can spot check the sizes.

What do you mean by spot check?  We need to validate the arguments to all
of our functions so we can throw exceptions with informative messages.

**Comment by bbbales2** [comment]
> So we also want to be able to append vector<Matrix<var,-1,-1.> and
> vector<Matrix<double,-1,-1>> as well as append vector<vector<var>> and vector<vector<double>>.
> There's no special logic for when the types happen to be the same.

Oh, arrays of arrays. I'd forgotten about that. How about I stop at the Eigen::Matrix, Eigen::Vector stuff? I'm not sure how to handle the std::vector<std::vector<...std::vector<T>>> situation elegantly (array of array of array ... where I don't know how deep it goes).

> What do you mean by spot check?

Good question. If either list is longer than length zero (so we're actually appending things) I check to make sure the shapes of the first element in each list are equal. I don't check them all against each other. I could easily do this (isn't going to cost much time), but it seemed more pedantic than useful (all Matrix/Vector/Rowvector types in a Stan list should be of a compatible size due to the declaration).

**Comment by bob-carpenter** [comment]
You can write a general converter using recursive template functions. But if want to stop at a limited set of signatures, that's OK.

Yes, you can assume all the data structures are square.    That will not be the case going forward, but makes sense now and the ragged types will be different in Stan if we ever get around to adding them.

**Comment by bbbales2** [comment]
Hopefully this should be good to go (pending builds).

Notes:
1. I had to modify the matrix version of append_array slightly to explicitly cast between Eigen matrices with different types: https://github.com/bbbales2/math/blob/771c9cb66bd6a73a91585a253ca98cbfb4c43005/stan/math/prim/mat/fun/append_array.hpp#L58 (http://stackoverflow.com/questions/23946658/error-mixing-types-with-eigen-matrices)

2. I added some metaprogramming stuff to the original version of append_array to make sure it's only working with scalar types (https://github.com/bbbales2/math/blob/771c9cb66bd6a73a91585a253ca98cbfb4c43005/stan/math/prim/mat/fun/append_array.hpp#L77). Without this the compiler would try to use this form of the function for mismatched Eigen types (for instance, a Matrix<double, Dynamic, Dynamic> and a Matrix<double, Dynamic, 1>). This is no good because there's no size checking in that version.

3. I did add some tests for vectors of matrices in all of prim/rev/fwd/mix. The tests where I check the elements aren't exhaustive, but I do think I at least test that all instantiations that make sense in Stan at compile without error.

4. I didn't do the recursive vector<vector<...>> stuff. I had a look around, but I wasn't sure how to do it.

**Comment by bob-carpenter** [comment]
I'll take a look.  Some initial comments below.

> On May 21, 2017, at 4:26 PM, Ben Bales <notifications@github.com> wrote:
> 
> Hopefully this should be good to go (pending builds).
> 
> Notes:
> 
> 	• I had to modify the matrix version of append_array slightly to explicitly cast between Eigen matrices with different types: https://github.com/bbbales2/math/blob/771c9cb66bd6a73a91585a253ca98cbfb4c43005/stan/math/prim/mat/fun/append_array.hpp#L58 (http://stackoverflow.com/questions/23946658/error-mixing-types-with-eigen-matrices)

These aren't technically casts in C++.  We run into this problem all over---last place I remember it coming up was for the conditional (ternary) operator; we also have something like this for assignment.  Feel free to write to discourse or an issue asking how we've solved problems like this before.

You can also see the parallel in our existing append_rows and append_cols, which also has to work on mixed type input. 

> 	• I added some metaprogramming stuff to the original version of append_array to make sure it's only working with scalar types (https://github.com/bbbales2/math/blob/771c9cb66bd6a73a91585a253ca98cbfb4c43005/stan/math/prim/mat/fun/append_array.hpp#L77). Without this the compiler would try to use this form of the function for mismatched Eigen types (for instance, a Matrix<double, Dynamic, Dynamic> and a Matrix<double, Dynamic, 1>). This is no good because there's no size checking in that version.

Cool.  I didn't know they had a .cast() method, though it's a very very confusing name as it doesn't do C++ casting.  So maybe it's good you set out on your own on this one, because we should probably be using theirs in places where I wrote my own version (it's a trivial function, but even so, better someone else tests and maintains it).  

We should only be "casting" up from a double to an autodiff type.

> 
> 	• I did add some tests for vectors of matrices in all of prim/rev/fwd/mix. The tests where I check the elements aren't exhaustive, but I do think I at least test that all instantiations that make sense in Stan at compile without error.
> 
> 	• I didn't do the recursive vector<vector<...>> stuff. I had a look around, but I wasn't sure how to do it.

The trick is to think of it as vector<T> and you just want to append a sequence of T.  The problem is that the T may themselves require the same kind of casting on type as you did for matrices.  We won't be able to add a matching .cast() method to std::vector, so for this, we had to write our own.  But that'll let you append var[] and double[], for example---you see that in our existing methods for appending

**Comment by bbbales2** [comment]
@seantalts Any chance you could take a look at this? It's ready to go. I'm around all day to make changes.

I saw someone using as a replacement "to_array_1d(append_row(to_vector(ts), to_vector(V)))" on the Discourse (http://discourse.mc-stan.org/t/bug-in-expose-stan-functions-for-integrate-ode-bdf/830).

**Comment by seantalts** [comment]
Yep! Will take a look.

File: stan/math/prim/mat/fun/append_array.hpp [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+     * @tparam T2 Scalar Type of second vector
+     * @param x First vector
+     * @param y Second vector
+     * @return A vector of x and y concatenated together (in that order)
+     */
+    template <typename T1, typename T2>
+    inline typename
+    boost::disable_if_c<is_vector_like<T1>::value || is_vector_like<T2>::value,
+      std::vector<typename
+                  boost::math::tools::promote_args<T1, T2>::type> >::type
```
Comment (seantalts) [inline stan/math/prim/mat/fun/append_array.hpp]: Why do you use `promote_args` here and `return_type` on the implementation above (line 44)?

**Comment by seantalts** [comment]
Looks good to me! I had a likely unimportant question in the PR. Thanks!

File: stan/math/prim/mat/fun/append_array.hpp (hunk 1 above)
Comment (bbbales2) [inline stan/math/prim/mat/fun/append_array.hpp]: Good point, I'll change to return_type so the code is shorter.

**Comment by bbbales2** [comment]
@seantalts I changed the return type from promote_args to return_type. Makes the code prettier. Should be good to go. Thanks!

**Comment by bob-carpenter** [comment]
Why is there a `disable_if_c` here?  It looks like it'll prevent appending arrays that contain things other than scalars---is there a separate implementation for those or is it just not allowed? 

```
inline typename
boost::disable_if_c<is_vector_like<T1>::value || is_vector_like<T2>::value,
std::vector<typename return_type<T1, T2>::type> >::type
append_array(const std::vector<T1>& x, const std::vector<T2>& y) {
...
```

The only reason I can think of where you would need a `disable_if`s is to control an otherwise ambiguous instantiation.

**Comment by bbbales2** [comment]
> Why is there a disable_if_c here? It looks like it'll prevent appending arrays that contain things other than scalars

That was my original goal. Here's the detailed breakdown:

The issue was that non-scalar types were using this version of the function when their specializations didn't exist. Types like `vector<vector<double> >`, for instance, would compile. Because there is no size checking in this version of append_array, I thought this is undesirable behavior.

I just went to test this explicitly though, and it appears there is a different behavior between `boost::math::tools::promote_args` and `return_type`. When I use the version with promote_args, things compile (allowing non-safe behavior). With return_type, the compile fails (which is what I wanted to achieve with the metaprogramming).

I think what's happening is `return_type<T1, T2>::type` is extracting the actual scalar type from T1/T2, meaning that:

```
return_type<vector<vector<double> >, vector<vector<double > > >::type
append_array(vector<vector<double> >, vector<vector<double > >)
```

Turns into

```
vector<double>
append_array(vector<vector<double> >, vector<vector<double > >)
```

The compile then breaks cause there's no way to cast vector<double>s to doubles (which needs to be done internally in the function).

Should I use my metaprogramming or should I lean on return_type like this? It's like it's achieving the right result for not exactly the right reason, but it is achieving the result.

(by the way, hold the pull request -- I left a promote_args in that I want to remove if this is acceptable -- with apologies to Travis and Jenkins)

**Comment by bbbales2** [comment]
Actually, am I thinking about this design wrong? Should the invalid use cases I described above be controlled at the Stan interface level? The Stan interface will limit what types get used. If I don't add the stuff to the function_signatures file, it won't be possible.

Am I trying to be too picky in only allowing certain types through at the C++/math library level?

**Comment by bob-carpenter** [comment]
> On Jun 10, 2017, at 1:12 AM, Ben Bales <notifications@github.com> wrote:
> 
> Actually, am I thinking about this design wrong? Should the invalid use cases I described above be controlled at the Stan interface level?

That's what we've been doing elsewhere, which is why I was asking.


> The Stan interface will limit what types get used. If I don't add the stuff to the function_signatures file, it won't be possible.

Exactly.  At least from the Stan language.


> Am I trying to be too picky in only allowing certain types through at the C++/math library level?

We haven't tried to do that elsewhere.  The general idea in C++ is that it's up client (code that calls the API) to send it instances that support all the relevant concepts.

**Comment by bbbales2** [comment]
Excellent! I removed that code.

With apologies to Jenkins and Travis, this should be ready to go now.

**Comment by seantalts** [comment]
Bob, the missing functionality was not something the compiler would catch -
if there's vectors of matrices of vectors, this code won't check that the
dimensions of the matrices and their constituents line up. Which might be
fine for the math library anyway. Wasn't sure if you caught that based on
your comment about C++ concepts. Anyway I think this version is also fine
so we can merge it once all the tests pass in my opinion.

**Comment by bob-carpenter** [comment]
@seantalts --- you mean `Eigen::Matrix<std::vector<double>, -1, -1>` or something like that?  I don't think we have to worry about what happens if people try to use `std::vector` scalars in their `Eigen::Matrix` types.  I wasn't even going to worry about instantiations that don't match what we use in the math library, but I could see how you might want to try to control that for a general-purpose math lib.  I'm OK with this either way.  We haven't tried at all in other functions to prohibit this.  Most of the uses would fail to be instantiated anyway.

**Comment by seantalts** [comment]
Sounds good. I'll merge this in.

**Pull request merged**